R/subset_draws.R

```diff
@@ -195,6 +160,114 @@ subset_dims <- function(x, ...) {
   eval2(call, dots)
 }
 
+# internal method for subset_draws
+.subset_draws <- function(x, ...) {
+  UseMethod(".subset_draws")
+}
+
+#' @export
+.subset_draws.draws_matrix <- function(x, draw = NULL, variable = NULL,
+                                       reserved = FALSE, ...) {
+  out <- subset_dims(x, draw, variable)
+  if (reserved && !is.null(variable)) {
+    new_vars <- variables(out, reserved = TRUE)
+    reserved_vars <- setdiff(reserved_variables(x), new_vars)
+    if (length(reserved_vars)) {
+      x_reserved <- subset_dims(x, draw, reserved_vars)
+      out <- cbind(out, x_reserved)
+      class(out) <- class(x)
+    }
+  }
+  out
+}
+
+#' @export
+.subset_draws.draws_array <- function(x, iteration = NULL, chain = NULL,
+                                      variable = NULL, reserved = FALSE, ...) {
+  out <- subset_dims(x, iteration, chain, variable)
+  if (reserved && !is.null(variable)) {
+    new_vars <- variables(out, reserved = TRUE)
+    reserved_vars <- setdiff(reserved_variables(x), new_vars)
+    if (length(reserved_vars)) {
+      x_reserved <- subset_dims(x, iteration, chain, reserved_vars)
+      out <- abind(out, x_reserved, along = 3L)
+      class(out) <- class(x)
+    }
+  }
+  out
+}
+
+#' @export
+.subset_draws.draws_df <- function(x, iteration = NULL, chain = NULL,
+                                   draw = NULL, variable = NULL,
+                                   unique = TRUE, reserved = FALSE, ...) {
+  if (!is.null(variable)) {
+    x <- x[, variable, reserved = TRUE]
+  }
+  if (!is.null(draw)) {
+    # each x$.draw is unique so using 'match' is valid here
+    x <- x[match(draw, x$.draw), ]
+    x$.draw <- repair_iteration_ids(x$.draw)
+    x$.iteration <- x$.draw
+  } else if (!is.null(chain) || !is.null(iteration)) {
+    if (unique) {
+      if (!is.null(chain)) {
+        x <- x[x$.chain %in% chain, ]
+      }
+      if (!is.null(iteration)) {
+        x <- x[x$.iteration %in% iteration, ]
+      }
+      x <- repair_draws(x, order = FALSE)
+    } else {
+      # non-unique subsetting is conceptually easier in 'draws_list' objects
+      # TODO: perform non-unique subsetting directly in the 'draws_df' object
```

Comment:
Instead of a TODO comment in the code should we open an issue for this?

Reply:
I made an issue (#92)